qutip/wigner.py

```diff
@@ -603,20 +633,86 @@ def qfunc(state, xvec, yvec, g=sqrt(2)):
+def _qfunc_pure(psi, alpha_mat, precompute=False):
     """
     Calculate the Q-function for a pure state.
+
+    If provided, precompute needs to be computed with
+    qfunc_amat(xvec, yvec, n, g), where xvec, yvec and g
+    need to be the same as for qfunc, and n = np.prod(state.shape) is the dim
+    of a pure state in the chosen system. This gives 3-10x speedup per call
     """
     n = np.prod(psi.shape)
     if isinstance(psi, Qobj):
         psi = psi.full().flatten()
     else:
         psi = psi.T
 
-    qmat = abs(polyval(fliplr([psi / sqrt(factorial(arange(n)))])[0],
-                       conjugate(alpha_mat))) ** 2
+    if isinstance(precompute, np.ndarray):
+        qmat = np.dot(precompute, psi)
+    else:
+        qmat = polyval(fliplr([psi / sqrt(factorial(arange(n)))])[0],
+                       conjugate(alpha_mat))
+
+    # faster than np.abs()**2 if len(xvec) >~ 10
+    qmat = qmat.real**2 + qmat.imag**2
+    if not isinstance(precompute, np.ndarray):
+        qmat *= exp(-abs(alpha_mat) ** 2)
+    return qmat / pi
+
+
+def qfunc_precompute(xvec, yvec, n, g=sqrt(2), max_memory=1024):
+    """Helper matrix for fast Q-function at points `xvec + i * yvec`.
+
+    Explanation: A lot of the cost of the Husimi Q function does not depend on
+    the state. If it is called many times (e.g. for a density matrix or when
+    doing multiple states), this can be used for a speedup. This function
+    precomputes everything that does not depend on the state and stores it as
+    a 3d array. The Q function itself is then a dot product between the last
+    axis of the precomputed array and a pure state.
+
+    Warning: The returned array has size
+             len(xvec) * len(yvec) * n * 16 Byte, can be large
+
+    Parameters
+    ----------
+    xvec : array_like
+        x-coordinates at which to calculate the Wigner function.
+
+    yvec : array_like
+        y-coordinates at which to calculate the Wigner function.
+
+    n : int
+        Dimension of the pure states of which the Q func will be evaluated
+
+    g : float
+        Scaling factor for `a = 0.5 * g * (x + iy)`, default `g = sqrt(2)`.
+
+    max_memory : float
+        Maximal memory size in MB, default 1GB.
+
+    Returns
+    --------
+    precomputed : array
+        Precomputed array that contains everything in _qfunc_pure that is not
+        dependent on the state.
+
+    """
+    memory = len(xvec) * len(yvec) * n * 16 / 1024**2
+    if memory > max_memory:
+        raise MemoryError(
+            f"Precomputation uses {memory} MB memory, with a max of "
+            + f"{max_memory} MB.Turn precomputation off with precompute=False "
+            + f"or use a larger max."
+        )
+    X, Y = meshgrid(xvec, yvec)
+    amat = 0.5 * g * (X - Y * 1j)
 
-    return real(qmat) * exp(-abs(alpha_mat) ** 2) / pi
+    powers = np.arange(n)
+    precomputed = np.power(np.expand_dims(amat, axis=-1), powers)
+    precomputed /= sqrt(factorial(arange(n)))
```

Comment:
Is this possibly dangerous?

It hits the limit of floating-point computation earlier than it needs to.  For one,
```python
np.sqrt(scipy.special.factorial(np.arange(n)))
```
is technically a quadratic dependency, but its run cost is completely overshadowed by the other stuff so it's not a big deal.  _Somewhat_ more importantly, it starts doing funny business at `n` is 171 due to the maximum factorial that can be represented by a double-precision float, even though we then take the square root.  We can extend the limit to 302 by instead doing
```python
precomputed[:, :, 2:] /= np.cumprod(np.sqrt(np.arange(2, n)))
```
That said, it's not _so_ important because we're using it as a divisor, and floating point multiplies overflow to `inf`, and it's defined for all cases that `x / inf == 0.0` except for `x` is `inf` itself.  This could become a problem if `xvec` and `yvec` are sufficiently wide, though - it's possible for the powers call to overflow and produce `inf`, then get divided by an `inf`, even though the true result is representable by floating-point calculations.

I'm not certain that this is very important (presumably no-one uses values of `n` that large), but I saw a value of 256 in a test, and the current form technically can overflow that.  Just wanted to point it out.